### ISSUE TITLE
RollingSync functionality is vulnerable to a race condition around application refreshes

### ISSUE THREAD

**Issue by jutley** [argoproj/argo-cd#12325]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

When defining an ApplicationSet with a rollout strategy as `RollingSync` and multiple steps, step dependencies are not respected when Applications are refreshed out-of-order.

My understanding is that for each ApplicationSet, if any Application is out-of-sync, then the Applications will get synced according to the RollingSync steps. However, only out-of-sync Applications are considered, and there is no effort to refresh all the Applications. This means that Applications in late steps can get synced _before_ the Applications in early steps.

I suspect this behavior can be resolved by triggering a refresh of all Applications in an ApplicationSet before attempting the RollingSync.

**To Reproduce**

Create an ApplicationSet with `RollingSync`, using two steps targeting its own Application. Do not use a webhook for triggering refreshes so that we can do it manually.

Push a change to git which will eventually cause both Applications in the ApplicationSet to go out of sync. Manually refresh the Application in the _second_ step before the Application in the first step refreshes. The Application we refresh will automatically sync despite the previous steps not syncing.

**Expected behavior**

The manually-refreshed Application will not sync until the Applications in the previous steps are refreshed and (if necessary) synced.

**Version**

```json
{
    "Version": "v2.6.0+acc554f",
    "BuildDate": "2023-02-06T21:17:26Z",
    "GitCommit": "acc554f3d99010e0353b498a595844b30090556f",
    "GitTreeState": "clean",
    "GoVersion": "go1.18.10",
    "Compiler": "gc",
    "Platform": "linux/amd64",
    "KustomizeVersion": "v4.5.7 2022-08-02T16:35:54Z",
    "HelmVersion": "v3.10.3+g835b733",
    "KubectlVersion": "v0.24.2",
    "JsonnetVersion": "v0.19.1"
}
```

**Comment by riuvshyn** [comment]
@wmgroot Hello, any chance you can look at this? 🙏🏽 

I was observing potentially similar issue, in my case I am  pushing chart version update to applicationSet which is not yet available which puts `ALL` applications in to `Unknown` status, eventually that chart version becomes available and every application will transition state from `Unknown` -> `OutOfSync` in pretty much random order which will cause Sync operation to start in not expected order. And even more most of times some Applications are left in `OutOfSync` state forever, I noticed that when I trigger manual Sync for application which is already Synced and Healthy this will unblock rolling Sync to proceed to `OutOfSync` Applications.

So I would agree with @jutley that it makes sense to perform a hard refresh on **ALL** applications for given ApplicationSet before starting rolling sync.

**Comment by riuvshyn** [comment]
still actual with 2.7.10

**Comment by riuvshyn** [comment]
rollingsync is still very unreliable on 2.8.0

**Comment by crenshaw-dev** [comment]
@riuvshyn for the purposes of this thread, I think we can assume that the problem persists as long as no one has responded with a solution.

**Comment by aliazlan-t** [comment]
same issue in v2.10.2

**Comment by jiaxwu** [comment]
> I suspect this behavior can be resolved by triggering a refresh of all Applications in an ApplicationSet before attempting the RollingSync.
I believe this change is worth making, as I have encountered the same issue.

**Comment by fidelcanojr** [comment]
If your target revision is a branch, I've found a workaround. I'll illustrate this with GitHub as the source provider, but a similar approach should work for other git providers. Say you have an application set with one generator:

```
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
spec:
    generators:
        - your-generator: {}
    template:
        spec:
            source:
                repoUrl: https://github.com/your-org/your-repo.git
                targetRevision: your-branch
```

You can use the SCM generator to filter your branch and map it to a sha and use that as your target revision instead. Merging this with your current generator using a matrix generator would look something like:

```
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
spec:
    generators:
      - matrix:
             generators:
                - your-generator: {}
                - scmProvider:
                      cloneProtocol: https
                      github:
                          organization: your-org
                          allBranches: true
                      filters:
                          - repositoryMatch: ^your-repo$
                            branchMatch: ^your-branch$
    template:
        spec:
            source:
                repoUrl: {{url}}
                targetRevision: {{sha}}
```

The generated applications will have their target revisions pinned to a sha, so you don't have to worry about an application going out of sync before the application set that generated it begins the progressive rollout. Once the application set recognizes the change to the branch, it'll update the target revision in all applications and sync them in the expected order.

It would be nice to have this issue addressed though so the application set definition can remain source provider agnostic and you can do things like tag tracking.

**Comment by shaxbee** [comment]
@crenshaw-dev we're facing this issue as well, if there is no one working on the fix I can dedicate some time to investigate and raise PR, any pointers?

**Comment by andrey-parkhomets** [comment]
is there any woraround ?
